[REF] merge rainbow man in one svg and revamp scss

Brieuc-brd opened the pull request:
Prior this PR the rainbow man was built with many separate SVGs and a lot of raw SCSS.
This commit refactors the rainbow man in one SVG in order to use less raw SCSS and more SVG attributes instead.

It also improves the overall animation with a little rotation of the rainbow man's head during the fade in animation.

task-2742831

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/83403).

Brieuc-brd commented:
Hello @stefanorigano 👋 ,

To remove raw SCSS as much as possible I refactored the rainbow man in only one SVG, including the message box which is built in HTML.

To do that, I had to move this HTML part in a `foreignObject`. This action created a little positional issue in the shadow, below the message box, on Safari (`transform and `position` are not well supported in a `foreignObject` in this browser…).

So to fix that, I had no choice but to remove this shadow and create a new one outside the `foreignObject`. So now the shadow is fixed below the SVG and is not relative to the message box anymore.

I think this change is worth it for the number of lines removed.

![rainbow_bef_aft](https://user-images.githubusercontent.com/80679690/151179645-c5ab85d0-45e3-48f1-8cde-0b393404af6c.jpg)

Also to improve the render, I changed the color of the shadow and added an animation which rotate and scale up a bit the head during the fade in animation. It adds a couple of new SCSS line but it's still less than before.

What do you think ?

stefanorigano commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,156 @@
-  $reward-size-mobile: 300px;
-  $reward-text-color: #727880;
-  $reward-base-time: 1.4s;
+    $reward-base-time: 1.4s;
```
Prefix **local** variables with `-`, in this case `$-reward-base-time`

stefanorigano commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,156 @@
-  &.o_reward_fading {
-    display: block;
-    animation: reward-fading-reverse ($reward-base-time * 0.4) ease-in-out 0s 1 normal forwards;
+    inset: 0; // Same than : left: 0; right: 0; top: 0; bottom: 0;
```
This type of comments are unnecessary ;)

stefanorigano commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,156 @@
-      path {
-        animation: reward-rainbow-reverse ($reward-base-time * 0.5) ease-out 0s 1 normal forwards;
-      }
+    &.o_reward_fading {
```
Add an empty line before ` &.o_reward_fading`... the same for all the other selectors

stefanorigano commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,156 @@
-      @include size($reward-size-mobile * 0.75, $reward-size-mobile / 2);
+    .o_reward_rainbow_man,
+    foreignObject {
+        // Could be replaced by the "overflow-visible" class after the migration to BS5
```
define "overflow-visible" in `web\...\webclient.scss` directly.
Add a note saying that the class can be removed with v5 (like  "Inline Margins")

stefanorigano commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,156 @@
+    .o_reward_face {
+        animation: reward-float $reward-base-time ease-in-out $reward-base-time infinite alternate;
+    }
+    .o_reward_star_01,
```
Related selectors, both can stay in the same line 
https://cssguidelin.es/#anatomy-of-a-ruleset

stefanorigano reviewed (changes requested):
Perfect, thanks... some minor conventions remarks

Brieuc-brd commented:
> Perfect, thanks... some minor conventions remarks

@stefanorigano Done ! 👍

Brieuc-brd commented:
Hello @odoo/rd-security 👋 ,

Could you check the security warning ?
I didn’t change the block with the `t-raw` attribute but, I refactored the structure around it (put it into a `foreignObject` into the SVG).

MiquelRForgeFlow commented:
The stars got bigger 😆 ⭐️

Brieuc-brd commented:
> The stars got bigger 😆 ⭐️

In the previous image, yes... because the screenshots were not taken at the same time during the animation 😉 .
But the size of the element and the animation are the same as before.

xmo-odoo commented:
@robodoo override=ci/security

Just one question: is support fot HTML messages in rainbowman actually useful?

Brieuc-brd commented:
> @robodoo override=ci/security
> 
> Just one question: is support fot HTML messages in rainbowman actually useful?

Yes, it seems that most of the Rainbow man messages are using `<b>`, `<strong>` or `<i>` tags.

stefanorigano commented:
Perfect, thanks @Brieuc-brd , you can remove the draft state.
@adr-odoo , ready for tech. review

Brieuc-brd commented:
Hello @stefanorigano 👋 ,

To avoid some issues with the `foreignObject` (eg. with Safari), I updated the rainbow man to not use it anymore.

Brieuc-brd commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,164 @@
+    inset: 0;
+    will-change: transform;
+    z-index: $zindex-modal;
+    animation: reward-fading $-reward-base-time * 0.5 ease-in-out forwards;
```
I’ve made some tests, and it's possible but not for all animations.
The "reverse" animation should be in CSSS because they are managed by the JS.
They are also two animations related to HTML tags that should stay in css : `.o_reward` and `.o_reward_msg_container`.

At the end we will have a mix between CSS and SVG animations. The base time variable will be split into 2 variables, one in the CSS and the other in Qweb.

Is it worth for the code maintenance to split these animations ? What do you think ?

Brieuc-brd commented:
@stefanorigano 👋 
The PR is updated.

stefanorigano commented on addons/web/static/src/core/effects/rainbow_man.scss:
```diff
@@ -1,256 +1,158 @@
-  to {
-    transform: translateY(5px);
-  }
+    from {
```
I think you can remove `from { }` and define the last animation step only.
Could you check the other animations too?

> If a 0% or from keyframe is not specified, then the user agent constructs a 0% keyframe using the computed values of the properties being animated. If a 100% or to keyframe is not specified, then the user agent constructs a 100% keyframe using the computed values of the properties being animated.
https://www.w3.org/TR/css-animations-1/#keyframes

Brieuc-brd replied:
@stefanorigano 👋 
It's done 👍

Brieuc-brd commented on addons/web/static/src/webclient/webclient.scss:
```diff
@@ -120,6 +120,15 @@ div.o_boolean_toggle.custom-control.custom-checkbox {
+
+.overflow-visible {
+  overflow: visible !important;
+}
```
It's done 👍

Brieuc-brd commented:
@stefanorigano 👋 
The branch is updated 🙂

stefanorigano commented:
@Brieuc-brd , could you rebase? thanks

Brieuc-brd commented:
> @Brieuc-brd , could you rebase? thanks

@stefanorigano 👋 
It's done 👍

Brieuc-brd commented:
@stefanorigano Hello again 👋 😄

stefanorigano commented:
@Brieuc-brd , good for me (you can remove the draft state)  👍 

@pparidans , ready for tech. review

adr-odoo reviewed (approved):
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward
